File: frontend/src/documents/draft.ts
import { HEALTH_KIND_LABELS, type HealthKind } from '../types'

/**
 * What a document reader returns and what the health-log form is prefilled with. The user reviews
 * and edits it before anything is saved; the model's output is never stored unseen.
 */
export interface HealthDraft {
  date: string
  kind: HealthKind
  title: string
  body: string
}

export const DOCUMENT_HINTS = ['auto', 'lab', 'imaging', 'letter', 'medication', 'diagnosis'] as const
export type DocumentHint = (typeof DOCUMENT_HINTS)[number]

/** Instruction sent with the pages. Transcription only: no interpretation, no advice. */
export function documentPrompt(hint: DocumentHint): string {
  const what = hint === 'auto' ? 'a medical document' : `a ${HEALTH_KIND_LABELS[hint].toLowerCase()}`
  return [
    `The attached pages are ${what} belonging to the user. Transcribe it faithfully into the JSON schema.`,
    'date: the document date in YYYY-MM-DD, or "" if not printed. kind: one of lab, imaging, letter, medication, diagnosis, other.',
    'title: a short label (test panel, modality and body region, or letter subject).',
    "body: the findings and conclusion in the document's own words, translated to English if needed, preserving the original wording of the conclusion. Do not add interpretation, risk statements or advice.",
    'values: every measured value as {name, value, unit, ref} for lab sheets; empty otherwise. Keep names as printed.',
    'Leave out patient name, date of birth, address, insurance and record numbers.',
  ].join(' ')
}

/** Gemini `responseSchema` (OpenAPI subset). */
export const DOCUMENT_SCHEMA = {
  type: 'object',
  properties: {
    date: { type: 'string' },
    kind: { type: 'string', enum: ['lab', 'imaging', 'letter', 'medication', 'diagnosis', 'other'] },
    title: { type: 'string' },
    body: { type: 'string' },
    values: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          name: { type: 'string' },
          value: { type: 'string' },
          unit: { type: 'string' },
          ref: { type: 'string' },
        },
        required: ['name', 'value'],
      },
    },
  },
  required: ['date', 'kind', 'title', 'body', 'values'],
}

const KINDS = new Set(Object.keys(HEALTH_KIND_LABELS))

/** Parses the model's JSON into a draft; tolerant of missing fields, strict about the kind. */
export function draftFromJson(text: string, fallbackDate: string): HealthDraft {
  const j = JSON.parse(text) as Partial<{
    date: string
    kind: string
    title: string
    body: string
    values: { name: string; value: string; unit?: string; ref?: string }[]
  }>
  const values = (j.values ?? [])
    .filter((v) => v?.name && v.value !== undefined)
    .map((v) => `${v.name}: ${v.value}${v.unit ? ` ${v.unit}` : ''}${v.ref ? ` (ref ${v.ref})` : ''}`)
  const body = [j.body?.trim() ?? '', values.join('\n')].filter(Boolean).join('\n\n')
  return {
    date: /^\d{4}-\d{2}-\d{2}$/.test(j.date ?? '') ? (j.date as string) : fallbackDate,
    kind: KINDS.has(j.kind ?? '') ? (j.kind as HealthKind) : 'other',
    title: j.title?.trim() || 'Untitled document',
    body,
  }
}

File: mobile/android/app/src/main/kotlin/com/hearth/documents/Draft.kt
package com.hearth.documents

import com.hearth.data.HealthKind
import org.json.JSONArray
import org.json.JSONObject

/**
 * What the document reader returns and what the add form is prefilled with (documents/draft.ts).
 * The user reviews and edits it before anything is saved; model output is never stored unseen.
 */
data class HealthDraft(val date: String, val kind: HealthKind, val title: String, val body: String)

/** What the user says the document is; "auto" lets the model decide. */
val DOCUMENT_HINTS = listOf("auto", "lab", "imaging", "letter", "medication", "diagnosis")

/** The instruction sent with the pages. Transcription only: no interpretation, no advice. */
fun documentPrompt(hint: String): String {
    val what = if (hint == "auto") "a medical document" else "a ${HealthKind.of(hint).label.lowercase()}"
    return listOf(
        "The attached pages are $what belonging to the user. Transcribe it faithfully into the JSON schema.",
        "date: the document date in YYYY-MM-DD, or \"\" if not printed. kind: one of lab, imaging, letter, medication, diagnosis, other.",
        "title: a short label (test panel, modality and body region, or letter subject).",
        "body: the findings and conclusion in the document's own words, translated to English if needed, preserving the original wording of the conclusion. Do not add interpretation, risk statements or advice.",
        "values: every measured value as {name, value, unit, ref} for lab sheets; empty otherwise. Keep names as printed.",
        "Leave out patient name, date of birth, address, insurance and record numbers.",
    ).joinToString(" ")
}

/** Gemini `responseSchema` (an OpenAPI subset), the same object the web sends. */
fun documentSchema(): JSONObject {
    fun str() = JSONObject().put("type", "string")
    val value = JSONObject().put("type", "object")
        .put("properties", JSONObject().put("name", str()).put("value", str()).put("unit", str()).put("ref", str()))
        .put("required", JSONArray().put("name").put("value"))
    return JSONObject().put("type", "object")
        .put(
            "properties",
            JSONObject()
                .put("date", str())
                .put("kind", str().put("enum", JSONArray(listOf("lab", "imaging", "letter", "medication", "diagnosis", "other"))))
                .put("title", str())
                .put("body", str())
                .put("values", JSONObject().put("type", "array").put("items", value)),
        )
        .put("required", JSONArray(listOf("date", "kind", "title", "body", "values")))
}

private val DATE = Regex("^\\d{4}-\\d{2}-\\d{2}$")

/** The model's JSON as a draft: tolerant of missing fields, strict about the kind and the date. */
fun draftFromJson(text: String, fallbackDate: String): HealthDraft {
    val j = JSONObject(text)
    val values = j.optJSONArray("values") ?: JSONArray()
    val lines = (0 until values.length()).mapNotNull { i ->
        val v = values.optJSONObject(i) ?: return@mapNotNull null
        val name = v.optString("name")
        if (name.isEmpty() || !v.has("value")) return@mapNotNull null
        val unit = v.optString("unit")
        val ref = v.optString("ref")
        "$name: ${v.optString("value")}${if (unit.isNotEmpty()) " $unit" else ""}${if (ref.isNotEmpty()) " (ref $ref)" else ""}"
    }
    val body = listOf(j.optString("body").trim(), lines.joinToString("\n")).filter { it.isNotEmpty() }.joinToString("\n\n")
    val date = j.optString("date")
    val kind = j.optString("kind")
    return HealthDraft(
        date = if (DATE.matches(date)) date else fallbackDate,
        kind = HealthKind.entries.firstOrNull { it.id == kind } ?: HealthKind.OTHER,
        title = j.optString("title").trim().ifEmpty { "Untitled document" },
        body = body,
    )
}
